types/src/auction/detail.rs

```diff
@@ -1,17 +1,169 @@
 use alloc::vec::Vec;
+use core::convert::TryInto;
 
 use num_rational::Ratio;
 
-use super::{
-    Auction, BidPurses, UnbondingPurse, UnbondingPurses, BID_PURSES_KEY, DEFAULT_UNBONDING_DELAY,
-    SYSTEM_ACCOUNT, UNBONDING_PURSES_KEY,
-};
 use crate::{
-    auction::{internal, MintProvider, RuntimeProvider, StorageProvider, SystemProvider},
+    auction::{
+        constants::*, Auction, BidPurses, Bids, DelegatorRewardMap, Delegators, EraId,
+        EraValidators, MintProvider, RuntimeProvider, SeigniorageRecipientsSnapshot,
+        StorageProvider, SystemProvider, UnbondingPurse, UnbondingPurses, ValidatorRewardMap,
+    },
+    bytesrepr::{FromBytes, ToBytes},
     system_contract_errors::auction::{Error, Result},
-    Key, PublicKey, URef, U512,
+    CLTyped, Key, PublicKey, URef, U512,
 };
 
+fn read_from<P, T>(provider: &mut P, name: &str) -> Result<T>
+where
+    P: StorageProvider + RuntimeProvider + ?Sized,
+    T: FromBytes + CLTyped,
+{
+    let key = provider.get_key(name).ok_or(Error::MissingKey)?;
+    let uref = key.into_uref().ok_or(Error::InvalidKeyVariant)?;
+    let value: T = provider.read(uref)?.ok_or(Error::MissingValue)?;
+    Ok(value)
+}
+
+fn write_to<P, T>(provider: &mut P, name: &str, value: T) -> Result<()>
+where
+    P: StorageProvider + RuntimeProvider + ?Sized,
+    T: ToBytes + CLTyped,
+{
+    let key = provider.get_key(name).ok_or(Error::MissingKey)?;
+    let uref = key.into_uref().ok_or(Error::InvalidKeyVariant)?;
+    provider.write(uref, value)?;
+    Ok(())
+}
+
+pub fn get_bids<P>(provider: &mut P) -> Result<Bids>
```

Comment:
I don't this was done in this PR, so probably out of scope, but I think we should prefer to not prefix getters with `get_`.

Reply:
Much agreed